hearth/lib/hearth/rules_engine.rb

```diff
@@ -0,0 +1,133 @@
+# frozen_string_literal: true
+
+require 'cgi'
+require 'ipaddr'
+require 'uri'
+
+require_relative 'structure'
+
+module Hearth
+  # Functions in the Smithy rules engine are named routines that
+  # operate on a finite set of specified inputs, returning an output.
+  # The rules engine has a set of included functions that can be
+  # invoked without additional dependencies, called the standard library.
+  # @api private
+  module RulesEngine
```

Comment:
Re: What files should the endpoint generator produce?

In spirit of following [Ruby Style Guide](https://github.com/rubocop/ruby-style-guide?tab=readme-ov-file#one-class-per-file), I think it is worth exploring having separate files for each classes. A single file is manageable right now but could this grow in the future? If so, we could keep them separated.

Reply:
It depends. Sometimes anonymous/api private classes within another class or module are perfectly fine.